### DISCUSSION TITLE
Vote on new features!

### DISCUSSION THREAD

**Discussion by tianyu-l** [pytorch/torchtitan#693] [Ideas]
Hi torchtitanists,

Thank you for your interests in torchtitan!

Please upvote on what features you would like to see next, and add one if it's not already there. We'll try to prioritize on the most requested features.

**Comment by tianyu-l** [comment]
https://github.com/pytorch/torchtitan/issues/625 multimodal models

**Comment by tianyu-l** [comment]
https://github.com/pytorch/torchtitan/issues/184 MoE and Expert Parallel

**Reply by SeunghyunSEO** [reply]
dropless moe with expert parallelism without doubt.
it should be democratized more.

**Reply by samsja** [reply]
Yay very important imo

**Reply by mayank31398** [reply]
scattermoe might be useful?
https://arxiv.org/abs/2403.08245

**Reply by SeunghyunSEO** [reply]
to my best knowledge, scattermoe is compatible with FSDP (maybe 1) and IBM granite is trained with it successfully, but it does not support EP and FSDP2, right?
i want torch native EP haha

**Reply by mayank31398** [reply]
correct, we train Granite with scattermoe.
It works with FSDP-2 + TP as well.
this is the TP implementation: https://github.com/IBM/dolomite-engine/blob/main/dolomite_engine/hf_models/models/moe_dolomite_TP/moe_TP/scatter.py
this is the non-TP implementation: https://github.com/IBM/dolomite-engine/blob/main/dolomite_engine/hf_models/models/moe_dolomite/moe/scatter.py

**Reply by SeunghyunSEO** [reply]
> correct, we train Granite with scattermoe.
> 
> It works with FSDP-2 + TP as well.
> 
> this is the TP implementation: https://github.com/IBM/dolomite-engine/blob/main/dolomite_engine/hf_models/models/moe_dolomite_TP/moe_TP/scatter.py
> 
> this is the non-TP implementation: https://github.com/IBM/dolomite-engine/blob/main/dolomite_engine/hf_models/models/moe_dolomite/moe/scatter.py

oh, thank you so much for sharing.
ill try scatter moe with torchtitan or lingua :)

**Reply by casper-hansen** [reply]
This fork has some code for "InternLM3 MoE" that implements a dropless MoE. It's not a ScatterMoE implementation, but does use Grouped GEMM.
https://github.com/zigzagcai/mytorchtitan

**Reply by casper-hansen** [reply]
I'm linking this issue here as it pertains to the overall features for MoE.
DeepSeek V3 Support: https://github.com/pytorch/torchtitan/issues/760

**Comment by tianyu-l** [comment]
state space models (e.g. Mamba)

**Comment by tianyu-l** [comment]
diffusion models

**Reply by giladturok** [reply]
this!

**Reply by tangjiasheng** [reply]
maybe this need to open a new repo.

**Reply by tianyu-l** [reply]
added here https://github.com/pytorch/torchtitan/tree/main/torchtitan/experiments/flux

**Comment by ad8e** [comment]
What about varlen ring attention? There are a few variants which are not stable enough for production use; its engineering/testing obstacles make it suitable for the strengths of the PyTorch folks.

**Reply by janEbert** [reply]
Should now be implemented by this, right?
https://discuss.pytorch.org/t/distributed-w-torchtitan-breaking-barriers-training-long-context-llms-with-1m-sequence-length-in-pytorch-using-context-parallel/215082

**Reply by tianyu-l** [reply]
added by https://github.com/pytorch/torchtitan/pull/1001
Context Parallel support is ongoing

**Comment by casper-hansen** [comment]
Evaluations implementation with support for 4D parallelism

**Comment by tangjiasheng** [comment]
I'd like to see a Deepspeed-Ulysses-Style sequence parallel used with DTensor.

**Comment by casper-hansen** [comment]
Qwen 2.5 support. Features needed on top of `llama.py`:
- `safetensors` to DCP
- attention bias on QKV
- [YaRN](https://arxiv.org/abs/2309.00071) (implemented in `qwen2.py` in HF transformers)

**Reply by conceptofmind** [reply]
Definitely should add YaRN

**Comment by dz1iang** [comment]
Parallel training is non-intrusive to model implementation.

**Reply by tianyu-l** [reply]
Can you elaborate more on this request? I believe our current parallelisms have largely avoided intrusive changes to model implementations.

**Reply by dz1iang** [reply]
> Can you elaborate more on this request? I believe our current parallelisms have largely avoided intrusive changes to model implementations.

I have an idea to modify parallelize_llama.py into parallelize_model.py, allowing for quick parallel training of HuggingFace models.

**Comment by samsja** [comment]
Would be nice to have a native (and tested) transformer export script as well. There is some trick to do with the complex number rope implementation in the conversion. We have a script to do this that should be compatible with the torchtitan implementation as it was originally copied from it : https://github.com/PrimeIntellect-ai/prime/blob/main/scripts/export_dcp.py

**Comment by samsja** [comment]
Would be nice as well to have an implementation of sequence packing / document masking using flex attention like in torchtune

**Reply by tianyu-l** [reply]
added by https://github.com/pytorch/torchtitan/pull/1001

**Reply by samsja** [reply]
thanks !!!

**Comment by functionstackx** [comment]
It seems like the biggest thing missing from torchtitan that most production gaint LLM training runs have is fault tolerance training

I.e. can quickly recover from a failure by removing just 1 data replicas instead of stopping the whole workload and restarting from ckpt

**Reply by zhangmozhe** [reply]
This is indeed useful

**Reply by tianyu-l** [reply]
added in https://github.com/pytorch/torchtitan/pull/834

**Comment by cailun01** [comment]
+ Training data preprocessing examples.
+ Multimodal model training.

**Comment by eminorhan** [comment]
- A complete example script demonstrating end-to-end evaluation of saved checkpoints (e.g. using `torchtune` and `lm-eval`).

**Reply by casper-hansen** [reply]
I already put this as a comment and I agree it would be nice if we can use something standardized like you mentioned.
https://github.com/pytorch/torchtitan/discussions/693#discussioncomment-11355716

**Comment by eminorhan** [comment]
- Handling more complicated datasets (e.g. interleaved datasets with given mixing probabilities, support for shuffling, etc.).

**Reply by samsja** [reply]
if it can help I have a simple interleaved / mixing probability dataset implementation here : https://github.com/PrimeIntellect-ai/prime/blob/main/src/zeroband/data.py

**Comment by zhangmozhe** [comment]
Reinforcement learning with verifiable rewards as used in TULU3 and DeepSeek R1, with asynchronous RL training

**Reply by SeunghyunSEO** [reply]
bro torchtitan is not for post-training.
torchtitan aims to showcase the cutting-edge large scale distributed training features of the latest pytorch in a clean, minimal codebase.
you should request RLVR feature at [torchtune](https://github.com/pytorch/torchtune)

**Comment by janEbert** [comment]
Post-training conversion to HuggingFace.

This is quite the beast due to the various options that TorchTitan offers, and not all of them being part of the model code (e.g., FP8 conversion, CP requiring an external context manager, etc.). All that combined with the maintenance aspects of future changes.
My current method for achieving at least evaluation support with an HF-compliant `forward` front-end is a client-server implementation, in which a distributed TorchTitan model (which is entirely and only dependent on TorchTitan code), is queried from a non-distributed HF model. However, this doesn't handle any of the post-training use cases and isn't really in line with HF's APIs; however, this implementation minimizes maintenance due to adapting most TorchTitan changes automatically.

**Reply by samsja** [reply]
evaluation via hf is quite important imo, as most of the eval code are compatible out of the box with hf api

**Reply by tianyu-l** [reply]
@samsja I'm still learning the eval area and have some questions.

I wonder why "eval via hf" is important, compared with just "eval via lm_eval". E.g. it seems as long as we inherit `lm_eval` base class and implement the key functions (according to https://github.com/EleutherAI/lm-evaluation-harness/blob/main/docs/model_guide.md#interface), we should be able to enjoy all the datasets & metrics in `lm_eval`.

Also wouldn't HF model put restrictions on the sets of parallelisms a model can run -- e.g. can HF model run Pipeline Parallel / Context Parallel?

**Reply by samsja** [reply]
hey @tianyu-l 

Yes eval via lm_eval would work as well. 

Also HF implementation is the default implementation for llama code, and any model trained with torchtitan will likely be released and eval via hf impl alter on. Therefore, I believe that TorchTitan should maintain an export function to hf safetensors format. It would also make model train with torchtitan compatible with a broader set of open source tool including vllm, sglang etcc ...

> Also wouldn't HF model put restrictions on the sets of parallelisms a model can run -- e.g. can HF model run Pipeline Parallel / Context Parallel?

I think that for eval TP / DP is enough at least up to 70B model on 8H100

**Reply by janEbert** [reply]
The reason for me to have HF compatibility is because there are several other frameworks, such as the excellent [EuroEval suite](https://github.com/EuroEval/EuroEval), which potentially include evaluation tasks that are not available in `lm_eval`. :)
It is simply a more general API for the world of evaluation, but also enables additional features for free, like chat deployment.

**Reply by tianyu-l** [reply]
@samsja @janEbert 
It sounds that instead of building eval directly in torchtitan, converting to HF format and doing downstream work with HF model definition is more preferable. If that's the case we already have DCP -> .pt -> HF (https://github.com/pytorch/torchtitan/issues/758, https://github.com/pytorch/torchtitan/issues/420); so maybe we should support direct DCP -> HF conversion?

This is in contrast to the request https://github.com/pytorch/torchtitan/discussions/693#discussioncomment-11440140
cc @casper-hansen @eminorhan any opinions?

How about in-training eval requested here https://github.com/pytorch/torchtitan/issues/883#issuecomment-2847282588
Do people think this feature is useful or not?

**Reply by casper-hansen** [reply]
I think evals during training is highly valuable. Keep the format in torchtitan for all the parallelism available.

**Reply by eminorhan** [reply]
My current eval pipeline is `dcp` -> `.pt` and then `torchtune` of `.pt` (with `lm-eval`). That's just two lines total, so I haven't really felt a pressing need for in-training evals yet but I can see that it would be useful to utilize the parallelisms already set up during training to make evals more efficient, and in fact this may be necessary for very large models (but I haven't been there yet personally).

On the other hand, I feel like this feature (parallelized eval based on `dcp` checkpoint) should really be implemented in `torchtune` or in `HF` rather than in `torchtitan` though.
